Enable sorting by producer in shopfront

guidoDutra opened the pull request:
#### What? Why?

Closes #1861 

This will enable enterprises to toggle between sorting by category and sorting by producer in the shopfront

#### What should we test?

* in the shop preferences page, enterprise can toggle shopfront sorting method between "By category" and "By producer"
* shopfront should display items according to the selected sorting method (category or producer) and to the list of preferred categories / producers
* if the preferred list is empty for the selected method, shopfront items must be displayed alphabetically (see note below)

#### Release notes

Enabled ordering by producer in shopfront

Changelog Category: User facing changes

#### Notes

* Currently, if the list of list of categories / producers is empty (see enterprise/{enterprise_name}/edit#/shop_preferences page), the shopfront items are sorted alphabetically per product name, regardless of producer or category. Should we instead consider sorting alphabetically per category name or producer name, whichever is selected in the shop_preferences page?

guidoDutra commented on app/assets/stylesheets/admin/disabled.scss:
```diff
@@ -16,4 +16,8 @@ input[type='button'], input[type='submit'] {
     color: #a1a1a1;
   }
+
+  .select2-search-choice {
+    background-color: #c3c3c3 !important;
+  }
```
I took the liberty of disabling the selector box if the ordering method is not selected (if "By producer" is selected, the "Category" selector is disabled). Then I added this class because the cards lables were not readable for the disabled box (white text on light gray background). Please let me know if I should declare this in a different way, or if the selector box should be always enabled for both methods

luisramos0 replied:
I think it's fine, I am just wondering if there is a better css way to do that without  !important.

jibees replied:
I maybe missed something, but I cannot find any `div` with class that is `select2-search-choice`. 
I just find this CSS rule that is used to gray the disabled field:
```css
.select2-container-multi.select2-container-disabled .select2-choices {
  background-color: #f4f4f4;
}
```

available in the `select2-rails` gem.

Could you please explain me which part of the part you want to disable with this rule?
Thanks a lot! 🙏

luisramos0 commented on app/services/products_renderer.rb:
```diff
@@ -59,7 +59,12 @@ def distributed_products
 
   def taxon_order
-    if distributor.preferred_shopfront_taxon_order.present?
+    if distributor.preferred_shopfront_product_sorting_method == "by_producer" && distributor.preferred_shopfront_producer_order.present?
```
this method is now the wrong name, you should rename it to products_order maybe because it can be taxons or producers now?

luisramos0 commented on app/assets/javascripts/admin/enterprises/directives/producer_autocomplete.js.coffee:
```diff
@@ -0,0 +1,30 @@
+        element.select2("container").find("ul.select2-choices").sortable
+          containment: 'parent'
+          start: -> element.select2("onSortStart")
+          update: -> element.select2("onSortEnd")
```
This is all copy pasted, can you please extract part of this code to a service and use it in both this ofnProducerAutocomplete and  ofnTaxonAutocomplete?
You can for example create a service like autocompleteSelect2 that returns everything in the setTimeout method here and takes a parameter callback and another one called something like list where you can inject the suppliers from this one.

jibees replied:
I agree with @luisramos0. Did you already work on it @guidoDutra ?

jibees replied:
Sorry, just saw after ;)

luisramos0 commented on app/assets/javascripts/admin/enterprises/controllers/enterprise_controller.js.coffee:
```diff
@@ -90,3 +92,5 @@ angular.module("admin.enterprises")
     $scope.translation = (key) ->
       t('js.admin.enterprises.form.images.' + key)
+
+    RequestMonitor.load $scope.suppliers = Enterprises.index(action: "visible", ams_prefix: "basic", "q[is_primary_producer_eq]": "true")
```
I dont think this is the correct way to fetch enterprises and I dont think this is the correct angular controler to load them.

I think the best pace to tackle this problem is here
app/assets/javascripts/darkswarm/controllers/products_controller.js.coffee
where the products are fetched and the filters applied.

In terms of how to load the producers, you already have $scope.order_cycle that maybe already has some form of $scope.order_cycle.suppliers list , otherwise you could build it here in a similar way to how $scope.supplied_taxons is built.
another option could be to simply use  $scope.Products and fetch the suppliers of each product.

But wait! Why do you need the producers here??? The ordering should all be done in the backend, you are already doing it in taxon_order in the renderer... can you explain what needs to be done here?

guidoDutra replied:
The reasoning for this was to make the list of producers available to `producer_autocomplete`, so a list shows up when the user clicks the selector on the shop preferences page.

Initially I just built a list of all producers, but that should be filtered to get only relevant producers (it was suggested to get all producers included in the current order cycle, but I think there can be multiple open order cycles, so maybe we should get all producers included in any open order cycle).

But I do feel like this in not an appropriate way of doing that, I'm just trying to figure out how it should be done instead. If you have some tips please let me know :)

luisramos0 replied:
oh, we are in the backoffice here :see_no_evil:  sorry, I missed that.

It would be incorrect to use OCs here imo, I think you actually doing it right, are you using 
the visible option here: app/controllers/admin/enterprises_controller.rb:168 ?

that should be the correct list of enterprises to show here filtered by is_producer :+1: 

And I think this is actually the right place to do this. So, is there something not working?

guidoDutra replied:
Hey Luis, thanks, I was actually testing incorrectly, it seems like the the producer are being filetered correctly according to permissions! I just have a question regarding this (/app/assets/javascripts/admin/enterprises/controllers/enterprise_controller.js.coffee:96):
```
RequestMonitor.load $scope.suppliers = Enterprises.index(action: "visible", ams_prefix: "basic", "q[is_primary_producer_eq]": "true")
```
is the "visible" option you mentioned above what is being refered to by the `action: "visible"` param? Removing this param didn't seem to change the behavor (I still got the filtered producers), so what does it do? Also it seems like `ams_prefix` defaults to `"basic"`, so it could be removed too. Should I remove both parameters?
```
RequestMonitor.load $scope.suppliers = Enterprises.index("q[is_primary_producer_eq]": "true")
```

luisramos0 reviewed (commented):
Hello Guido! Congrats on this very good effort!!!! Really well done for a OFN starter :medal_military: 

I left a few comments and questions. I hope it's useful.

luisramos0 reviewed (commented):
Hello Guido! Congrats on this very good effort!!!! Really well done for a OFN starter :medal_military: 

I left a few comments and questions. I hope it's useful.

guidoDutra commented:
Hey @luisramos0 , just opened this PR. I couldn't find an alternative for the `!important` css fix. Codeclimate is also complaining about the hard-coded color value, but those are being used frequently on the file, so I left it as is. Please let me know if you have any ideas

jibees reviewed (commented):
Thanks a lot @guidoDutra for this one!
I let some comments, and more globally I think you should rewrite git history by squashing commit f8b5b9baef9888efed883e2e94b81aebe1a8b6ed and 27e00ca8f2cbfb35a0f862c53f067f0aa57a832f

guidoDutra commented:
> Thanks a lot @guidoDutra for this one!
> I let some comments, and more globally:
> 
> * I think you should rewrite git history by squashing commit [f8b5b9b](https://github.com/openfoodfoundation/openfoodnetwork/commit/f8b5b9baef9888efed883e2e94b81aebe1a8b6ed) and [27e00ca](https://github.com/openfoodfoundation/openfoodnetwork/commit/27e00ca8f2cbfb35a0f862c53f067f0aa57a832f)
> * I saw you also fixed to codeclimate complains: thanks!

Hey @jibees, thanks for the feedback, I'll squash the commits in a sec. Regarding the CSS issue: Its applicable for the category/producer selector textboxes in the shop preferences page (admin/enterprises/form/_shop_preferences:26). The issue is that the background and the individual selected items have the same background color, which is a light gray that also makes it hard to read the white text in the items, so I overwrote the rule to give the items a darker color. Here is how it looks before and after the chages:

![image](https://user-images.githubusercontent.com/16342917/122096219-b05c2b00-cde4-11eb-80ff-2dc209a09591.png)


![image](https://user-images.githubusercontent.com/16342917/122096156-a1757880-cde4-11eb-82b9-b193f880749f.png)

Please let me know if there is a better approach for dealing with this

jibees commented:
@guidoDutra 
Thanks !
To avoid `!important` rule (which is really unwanted ;)) I guess something like 
```diff
-
-  .select2-search-choice {
-    background-color: $disabled-background !important;
+  &.select2-container-multi .select2-choices .select2-search-choice {
+    background-color: $disabled-background;
```
is good enough. What to you think?

guidoDutra commented:
> @guidoDutra
> Thanks !
> To avoid `!important` rule (which is really unwanted ;)) I guess something like
> 
> ```diff
> -
> -  .select2-search-choice {
> -    background-color: $disabled-background !important;
> +  &.select2-container-multi .select2-choices .select2-search-choice {
> +    background-color: $disabled-background;
> ```
> 
> is good enough. What to you think?

Thanks @jibees, that works! Just pushed another commit with the change. Should I also squash this into the previous one?

jibees commented:
@guidoDutra

> Thanks @jibees, that works! Just pushed another commit with the change. Should I also squash this into the previous one?

Yes please, it's important for a great git history. Again thanks @guidoDutra !

jibees commented:
@guidoDutra 
I thinks this commit 85e723f5d5b4932aa52d8f3bbb4ce24644077d22 should be edited: squashed with previous or maybe simply edited its title/description. The title isn't enough meaningful. You can see our documentation about how to write a great commit: https://github.com/openfoodfoundation/openfoodnetwork/wiki/Making-a-great-commit

 What do you think?

guidoDutra commented:
> @guidoDutra
> I thinks this commit [85e723f](https://github.com/openfoodfoundation/openfoodnetwork/commit/85e723f5d5b4932aa52d8f3bbb4ce24644077d22) should be edited: squashed with previous or maybe simply edited its title/description. The title isn't enough meaningful. You can see our documentation about how to write a great commit: https://github.com/openfoodfoundation/openfoodnetwork/wiki/Making-a-great-commit
> 
> What do you think?

Hey @jibees, I decided to keep the commit just to make it easier to understand, so I renamed it to something more specific. I had looked at the "making a great commit" page before, but I wasn't sure how to deal with refactoring commits such as this. How do you usually handle those? I ask because sometimes there are several small code style changes, so they can be hard to name

jibees commented:
@guidoDutra 
> how to deal with refactoring commits such as this. How do you usually handle those?

Most of the time, I squashed them into the previous commit that introduce that need of refactoring. Thus, we have a clean git history, which is "_perfect_" at first read.

jibees reviewed (approved):
LGTM 👌

andrewpbrett reviewed (approved):
This looks good, I'm going to move it to Test Ready. 

@guidoDutra what do you think about adding a feature spec that checks that the actual shopfront sorts in the correct order? Or do you think the unit tests are sufficient here?

guidoDutra commented:
> This looks good, I'm going to move it to Test Ready.
> 
> @guidoDutra what do you think about adding a feature spec that checks that the actual shopfront sorts in the correct order? Or do you think the unit tests are sufficient here?

Hey @andrewpbrett, I think this is sufficiently tested in `products_renderer_spec.rb`, but let me know if you think its best to add more tests

filipefurtad0 commented:
Hey @guidoDutra,

Thanks for tackling this one.

This PR introduces the `Display ordering in shopfront`, in the Shop Preferences (side menu). Some test-cases are displayed on the video:

- Adding a Category
- Sorting by one Category
- Adding another Category and inverting order

- Adding an Enterprise the hub has authorization to add products to OC
- Adding other Enterprises and inverting the ordering

- Combining these actions with shopfront filters

![Peek 2021-07-06 15-24](https://user-images.githubusercontent.com/49817236/124617711-103b7400-de6f-11eb-9330-650c70a81ef8.gif)

Other test cases:
- adding an Enterprise to enable sorting only works if enterprise permissions are set (in `admin/enterprise_relationships`) to:
  - `to add to order cycle` or `Everything` option is selected
  -  it is not possible to if permissions are set to any of the other options (`to manage products`, `to edit profile`, `to add products to inventory`)
  - given the correct permissions, the enterprise can be added regardless of having any products added to the OC.

- if no Enterprise or Category is set then sorting is alphabetical, by product name

I think we're good to go!
Thanks for this awesome enhancement :tada: 

PS: one last question - I noticed you added the translation keys:
```
          display_ordering_in_shopfront: "Display ordering in shopfront:"
          shopfront_sort_by_category: "By category"
          shopfront_sort_by_producer: "By producer"

```

But I'm not sure these cover the default strings `Category` and `Producer` (see pic). I found these keys but under a different namespace, so I'm not sure they'll be picked up - see below (using FR locale):

![image](https://user-images.githubusercontent.com/49817236/124586902-eeca9000-de4e-11eb-8e09-67cac0b68760.png)

What do you think?

andrewpbrett commented:
I'll merge and the additional translations can be a separate PR

guidoDutra commented:
Hey @filipefurtad0, you're right, I forgot to add translation keys for the placeholders. I'll open a new PR for that as advided by @andrewpbrett

guidoDutra commented:
Here is the PR with the translation tags @filipefurtad0 @andrewpbrett: https://github.com/openfoodfoundation/openfoodnetwork/pull/7893
